meson.build

```diff
@@ -0,0 +1,86 @@
+# CONTACT: <mailto:michael@squidfarts.com>                                        #
+#                                                                                 #
+# NOTICES:                                                                        #
+#                                                                                 #
+# MIT License                                                                     #
+#                                                                                 #
+# Copyright (c) 2019 Team Squidfarts                                              #
+#                                                                                 #
+# Permission is hereby granted, free of charge, to any person obtaining a copy    #
+# of this software and associated documentation files (the "Software"), to deal   #
+# in the Software without restriction, including without limitation the rights    #
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell       #
+# copies of the Software, and to permit persons to whom the Software is           #
+# furnished to do so, subject to the following conditions:                        #
+#                                                                                 #
+# The above copyright notice and this permission notice shall be included in all  #
+# copies or substantial portions of the Software.                                 #
+#                                                                                 #
+# DISCLAIMERS:                                                                    #
+#                                                                                 #
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR      #
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,        #
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE     #
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER          #
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   #
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE   #
+# SOFTWARE.                                                                       #
+#                                                                                 #
+###################################################################################
+
+
+
+project('facil.io', 'c',
+    version         : '0.2.4',
+    license         : 'MIT',
+    meson_version   : '>=0.50.0',
+    default_options : 
+    [ 
+        'werror=true',
+        'optimization=3', 
+        'warning_level=3', 
+        'b_sanitize=undefined', 
+        'b_lto=true', 
+        'b_lundef=true' 
+    ]
+)
+cc = meson.get_compiler('c')
+args_for_langs = 'c'
+
+
+
+threads_dep = dependency('threads')
+m_dep = cc.find_library('m', required : false)
```

Comment:
Could we do something similar to his for OpenSSL?

In pseudo code (probably not valid mason):

```
m_dep = cc.find_library('crypt', required : false, defines: "HAVE_OPENSSL")
```

...?

Reply:
I believe we can do something similar.  I have to run some experiments when I get back.

Reply:
How about this block of Meson script?

# meson.build experiment for OpenSSL.
`
# *** Checks for OpenSSL    ***
openssl_option = get_option('openssl')
if openssl_option.disabled()
  openssl_dep = []
else
  # XXX: https://github.com/mesonbuild/meson/issues/2945
  openssl_dep = dependency('openssl', required: false)
  if openssl_dep.found()
    backends += ['openssl']
  else
    # MSVC builds of OpenSSL does not generate pkg-config files,
    # so we check for it manually here in this case, if we can't find those files
    # Based on the CMake check for OpenSSL in CURL's CMakeLists.txt,
    # on which headers we should check for
    openssl_headers = []
    foreach h : ['crypto.h', 'engine.h', 'err.h', 'pem.h',
                 'rsa.h', 'ssl.h', 'x509.h', 'rand.h', 'tls1.h']
      openssl_headers += 'openssl/' + h
    endforeach

    # OpenSSL 1.1.x and 1.0.x (or earlier) have different .lib names,
    # so we need to look for the correct pair

    # Find either libcrypto.lib (1.1.x) or libeay32.lib (1.0.x or earlier) first
    libcrypto_dep = cc.find_library('crypto', required: false)
    if libcrypto_dep.found()
      libssl = 'ssl'
    else
      libcrypto_dep = cc.find_library('eay32', required: openssl_option)
      libssl = 'ssleay32'
    endif

    if libcrypto_dep.found()
      # Find the corresponding SSL library depending on which crypto .lib we found
      libssl_dep = cc.find_library(libssl, required: openssl_option, has_headers: openssl_headers)
    endif

    if libcrypto_dep.found() and libssl_dep.found()
      openssl_dep = [libcrypto_dep, libssl_dep]
      backends += ['openssl']
    endif
  endif
endif
`

# meson_options.txt

`option('openssl', type: 'feature', value: 'disabled', description: 'support for OpenSSL networking configration')`

Reply:
Looks interesting but also very complex. Maybe we should drop the OpenSSL support for now and manage it the same way we will manage file names (maybe using `make` to auto-generate a new mason.build file)...?

Note that facil.io supports OpenSSL 1.1.x or higher. Lower versions aren't supported. This might cleanup some of the suggested script.

Reply:
I think this can be a lot shorter if I used Meson's CMake module [here](https://mesonbuild.com/CMake-module.html).

This version of [OpenSSL](https://github.com/janbar/openssl-cmake/blob/master/CMakeLists.txt) has a CMakeLists.txt. If it's ok with you then I can add less Meson code than I presented in my latest experiment.

Just a thought.

Reply:
This would result with something like this for example.

`cmake_mod = import('cmake')

#Configure the CMake project
sub_proj = cmake_mod.subproject('libsimple_cmake')

#Fetch the dependency object
cm_lib = sub_proj.dependency('cm_lib')

executable(exe1, ['sources'], dependencies: [cm_lib])`

Just like that.

Reply:
I trust you, since I know very little about Mason, but I want to make sure that we're not forcing Mason users to install CMake.

Does this change requires people to install cmake, or does mason parses OpenSSL's CMake file?

Reply:
Sadly yes, the CMake module route would require that the user would have to install CMake.  That’s why I don’t use it much.

Reply:
I think there may be a way to simply the previous experiment.  I going to try some things.

Reply:
This maybe better to use for OpenSSL support, what do you think?

`# MSVC builds of OpenSSL does not generate pkg-config files,
# so we check for it manually here in this case, if we can't find those files
openssl = dependency('openssl', required: cc.get_id() != 'msvc')

if not openssl.found()
  # OpenSSL 1.1.x and 1.0.x (or earlier) have different .lib names,
  # so we need to look for the correct pair

  # Find either libcrypto.lib (1.1.x) or libeay32.lib (1.0.x or earlier) first
  libcrypto = cc.find_library('libcrypto', required: false)
  if not libcrypto.found()
    libeay = cc.find_library('libeay32')
  endif

  # Find the corresponding SSL library depending on which crypto .lib we found
  if libcrypto.found()
    libssl = cc.find_library('libssl')
  elif libeay.found()
    ssleay = cc.find_library('ssleay32')
  endif

  # Now set the openssl dependency to the corresponding pair of .lib files that we found
  if libcrypto.found()
    openssl = [libcrypto, libssl]
  else
    openssl = [libeay, ssleay]
  endif
endif`

Reply:
Would that work?

Reply:
Sorry for the delay, my grandmother's in the hospital, so there's a bit of a time shortage.

The script looks great, though I didn't test it on my machine.

However, facil.io supports only OpenSSL 1.1.x or higher, so there's no need for the `libeay` part.

Lets fix that up and merge. I'll work on the rest of it later.

Reply:
If we drop support for versions less than 1.1.x, the resulting script would look something like this.

```
#MSVC builds of OpenSSL does not generate pkg-config files,
#so we check for it manually here in this case, if we can't find those files
openssl = dependency('openssl', required: cc.get_id() != 'msvc')

if not openssl.found()
        libcrypto = cc.find_library('libcrypto', required: false)
        libssl = cc.find_library('libssl')
        openssl = [libcrypto, libssl]
endif
```

Reply:
Sorry about your grandmother.

Reply:
Well OpenSSL is added.